### DISCUSSION TITLE
ESP8266+SX178  POCSAG,message length

### DISCUSSION THREAD

**Discussion by shitra99** [jgromes/RadioLib#1112] [Q&A]
hi, what is the limit on message length? I can't send more than 50 characters

**Comment by jgromes** [comment]
Interesting question - I don't actually know what is the maximum, I don't think the POCSAG specifies any limit. RadioLib too doesn't enforce any maximum length, but I will admit I have not tested any long messages.

**Reply by shitra99** [reply]
If you send 100 characters, 60 recived , if  send 200 characters the esp8266 reboots

**Reply by jgromes** [reply]
I was not able to replicate the reboot you reported. Since you have not posted any actual code or serial output, it is very hard to help you.

I did see that PDW was not able to print out more than 45 characters (as can be seen below), but I suspect it is due to a limitation of that program.

```
1234567  14:57:18 15-06-24 POCSAG-2 NUMERIC  1200   TONE ONLY
1234567  14:57:20 15-06-24 POCSAG-1 NUMERIC  1200   0123456789*U -[]

1234567  14:57:22 15-06-24 POCSAG-4  ALPHA   1200   012345678901234567890123456789012345678901234

1234567  14:57:28 15-06-24 POCSAG-2 NUMERIC  1200   TONE ONLY
1234567  14:57:30 15-06-24 POCSAG-1 NUMERIC  1200   0123456789*U -[]

1234567  14:57:32 15-06-24 POCSAG-4  ALPHA   1200   012345678901234567890123456789012345678901234

```

**Reply by shitra99** [reply]
45 characters is not a limitation of the program, try cap 100 and you will get 62 characters, cap 8 you will get 85 characters

**Comment by shitra99** [comment]
#include <RadioLib.h>


SX1278 radio = new Module(5, 4, 16, 15);



// create Pager client instance using the FSK module
PagerClient pager(&radio);

void setup() {
  Serial.begin(9600);

  // initialize SX1278 with default settings
  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK();

  // when using one of the non-LoRa modules
  // (RF69, CC1101, Si4432 etc.), use the basic begin() method
  // int state = radio.begin();

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize Pager client
  Serial.print(F("[Pager] Initializing ... "));
  // base (center) frequency:     434.0 MHz
  // speed:                       1200 bps
  state = pager.begin(159.200, 1200);
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }
}

void loop() {
  Serial.print(F("[Pager] Transmitting messages ... "));

  

  // finally, let's transmit an ASCII message now
  int state = pager.transmit("1234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890", 100, RADIOLIB_PAGER_ASCII);
  delay(500);

  // we can also send only a tone

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

  // wait for a second before transmitting again
  delay(60000);
}
![hackrf](https://github.com/jgromes/RadioLib/assets/153459056/e5172385-dc3d-4321-b0a7-2fb472077c74)
![pager](https://github.com/jgromes/RadioLib/assets/153459056/e1d20845-ae8f-4f18-a04f-d097b95b3006)

**Reply by shitra99** [reply]
sent 100 characters, received 62

**Reply by jgromes** [reply]
I don't have those devices, so I can't replicate this bhavior (as shown in my previous post). But this is an open-source project, so feel free to dig into the sources - pull requests containing bugfixes are always welcome.

**Reply by shitra99** [reply]
In transmission mode, need to connect pins DIO0 and DIO1? it works without them. And I found in the comments in file SX127x_FSK_Modem.ino "// NOTE: FSK modem maximum packet length is 63 bytes!", is this not the reason for the transmission limit of 62 characters?

**Reply by jgromes** [reply]
> In transmission mode, need to connect pins DIO0 and DIO1?

Strictly speaking, for POCSAG transmission you don't need these pins. However, for all otehr transmissions using SX127x, at least DIO0 is required.

> is this not the reason for the transmission limit of 62 characters

It is not, POCSAG transmission is not done using SX127x packet engine and FIFO which is where that 63 byte limitation comes from. In fact the 62 characters you are receiving do not correspond to 62 bytes being sent over the air, but much more due to encoding.

**Comment by shitra99** [comment]
there are inaccuracies when forming packets, when using CAP code 100, 62 characters are accepted, this is 2 batches. If you make CAP in the zero frame (for example 8) it will accept 23 characters more, part of the first batch is filled with empty words (ıdle words) so that CAP 100 gets into the 4th frame. i.e. in theory the encoder either inserts an ıdle word at the end of the second batch or does not start the 3rd batch with a sync word, if the CAP is 1234567, it recive
![photo_2024-12-05_21-44-38](https://github.com/user-attachments/assets/70e10030-5bbd-41de-a531-618f8dceda92)
 45 characters

**Reply by jgromes** [reply]
Finally had a bit time of time to take a look at this - it turns out you were correct, there was indeed a bug in the encoder that caused overwriting the sync words between batches. That's why the length of actually received data was dependent on the CAP code. It is fixed in master (01e7fa07312175fdd94b1c0544d5c8b82a0044e7) and will be released in the next version. With the fix, there should be no limit on the message size - well until you run out of memory anyway ... 160+ characters below.

```
  1234567  11:22:16 04-01-25 POCSAG-2 NUMERIC  1200   TONE ONLY
  1234567  11:22:18 04-01-25 POCSAG-1 NUMERIC  1200   0123456789*U -[]
 
  1234567  11:22:21 04-01-25 POCSAG-4  ALPHA   1200   Hello World!
                                                      01234567890123456789012345678901234567890123456789012345678901234567890123456
                                                      78901234567890123456789012345678901234567890123456789012345678901234567890123
                                                      456789
```
